icicle/include/icicle/program/program.h

```diff
@@ -0,0 +1,140 @@
+#pragma once
+
+#include <vector>
+#include <functional>
+#include "icicle/program/symbol.h"
+
+namespace icicle {
+
+  using InstructionType = uint32_t;
+
+  enum PreDefinedPrograms { AB_MINUS_C, EQ_X_AB_MINUS_C };
+
+  /**
+   * @brief A class that convert the function described by user into a program that can be executed
+   *
+   * This class recieves a Symbol instance that contains a DFG representing the required calculation.
+   * It generates a vector of instructions that represent the calculation.
+   * Each instruction has the following format.
+   * bits 7:0   - opcode according to enum OpCode
+   * bits 15:8  - operand 1 selector from the input vector
+   * bits 23:16 - operand 2 selector from the input vector
+   * bits 31:24 - result selector
+   */
+  template <typename S>
+  class Program
+  {
+  public:
+    // Generate a program based on a lambda function
+    Program(std::function<Symbol<S>(std::vector<Symbol<S>>&)> program_func, const int nof_inputs)
+    {
+      std::vector<Symbol<S>> program_inputs(nof_inputs);
+      set_as_inputs(program_inputs);
+      Symbol<S> result = program_func(program_inputs);
+      generate_program(result);
+    }
+
+    // Generate a program based on a PreDefinedPrograms
+    Program(PreDefinedPrograms pre_def)
+    {
+      switch (pre_def) {
+      case AB_MINUS_C:
+        m_nof_inputs = 3;
+        break;
+      case EQ_X_AB_MINUS_C:
+        m_nof_inputs = 4;
+        break;
+      default:
+        ICICLE_LOG_ERROR << "Illegal opcode: " << int(pre_def);
+      }
+      m_nof_outputs = 1;
+      int instruction = int(OpCode::NOF_OPERATIONS) + int(pre_def);
+      m_instructions.push_back(instruction);
+    }
+
+    // run over all symbols at the vector and set there gen process to OP_INPUT
+    void set_as_inputs(std::vector<Symbol<S>>& combine_inputs)
+    {
+      m_nof_inputs = combine_inputs.size();
+      for (int input_idx = 0; input_idx < m_nof_inputs; input_idx++) {
+        combine_inputs[input_idx].set_as_input(input_idx);
+      }
+    }
+
+    // run over the DFG held by result and gemerate the program
+    void generate_program(Symbol<S>& result)
```

Comment:
I think it would be better to remove this logic from this class.  I would make this class only a container for the compiled format.

In compilers like llvm they typically have passes on the IR (symbols graph here). So one pass will allocate constants for example, another will generate the code, another will print it to file etc. Then a passManager is able to execute all the passes in a correct order based on dependencies and finally last pass is code generation which should fill this program object. 
I think it's a better design be cause you generally don't want to modify this code and risk breaking it when adding new optimizations for example.

Reply:
also this aligns with the fact that you have m_mem_addr inside the symbol where technically the symbol has nothing to do with this. Ideally some pass needs to take care of it and store the mapping itself. Then the final code generation pass would use this information for generating the code.

Alternatively maybe you could add an attributes map to the Symbol object and store general things on it. Typically passes would write this and later passes read it. But anyway coupling symbol to the program implementation should be avoided.

Reply:
lets discuss